libyul/optimiser/ReasoningBasedSimplifier.cpp

```diff
@@ -0,0 +1,266 @@
+	m_solver->pop();
+	if (result == CheckResult::UNSATISFIABLE)
+		_if.condition = make_unique<yul::Expression>(Literal{locationOf(*_if.condition), LiteralKind::Number, "1"_yulstring, {}});
+
+	m_solver->push();
+	m_solver->addAssertion(condition != constantValue(0));
+	CheckResult result2 = m_solver->check({}).first;
+	m_solver->pop();
+	if (result2 == CheckResult::UNSATISFIABLE)
+	{
+		_if.condition = make_unique<yul::Expression>(Literal{locationOf(*_if.condition), LiteralKind::Number, "0"_yulstring, {}});
+		_if.body = yul::Block{};
+	}
+
+	m_solver->push();
+	m_solver->addAssertion(condition != constantValue(0));
+
+	ASTModifier::operator()(_if.body);
+
+	m_solver->pop();
+}
+
+ReasoningBasedSimplifier::ReasoningBasedSimplifier(
+	Dialect const& _dialect,
+	set<YulString> const& _ssaVariables
+):
+	m_dialect(_dialect),
+	m_ssaVariables(_ssaVariables)
+{
+	m_solver = make_unique<smtutil::SMTPortfolio>();
+}
+
+smtutil::Expression ReasoningBasedSimplifier::encodeExpression(yul::Expression const& _expression)
+{
+	return std::visit(GenericVisitor{
+		[&](FunctionCall const& _functionCall)
+		{
+			if (auto const* dialect = dynamic_cast<EVMDialect const*>(&m_dialect))
+				if (auto const* builtin = dialect->builtin(_functionCall.functionName.name))
+					if (builtin->instruction)
+						return encodeBuiltin(*builtin->instruction, _functionCall.arguments);
+			return newRestrictedVariable();
+		},
+		[&](Identifier const& _identifier)
+		{
+			if (
+				m_ssaVariables.count(_identifier.name) &&
+				m_variables.count(_identifier.name)
+			)
+				return m_variables.at(_identifier.name);
+			else
+				return newRestrictedVariable();
+		},
+		[&](Literal const& _literal)
+		{
+			smtutil::Expression v = valueOfLiteral(_literal);
+			if (!m_useInt)
+				v = int2bv(v);
+			return v;
+		}
+	}, _expression);
+}
+
+smtutil::Expression ReasoningBasedSimplifier::encodeBuiltin(
+	evmasm::Instruction _instruction,
+	vector<yul::Expression> const& _arguments
+)
+{
+	vector<smtutil::Expression> arguments = applyMap(
+		_arguments,
+		[this](yul::Expression const& _expr) { return encodeExpression(_expr); }
+	);
+	switch (_instruction)
+	{
+	case evmasm::Instruction::ADD:
+		return wrap(arguments.at(0) + arguments.at(1));
+	case evmasm::Instruction::MUL:
+		return wrap(arguments.at(0) * arguments.at(1));
+	case evmasm::Instruction::SUB:
+		return wrap(arguments.at(0) - arguments.at(1));
+	case evmasm::Instruction::DIV:
+		return smtutil::Expression::ite(
+			arguments.at(1) == constantValue(0),
+			constantValue(0),
+			arguments.at(0) / arguments.at(1)
+		);
+	// TODO SDIV
+	case evmasm::Instruction::MOD:
+		return smtutil::Expression::ite(
+			arguments.at(1) == constantValue(0),
+			constantValue(0),
+			// TODO compute mod differently?
+			wrap(arguments.at(0) % arguments.at(1))
+		);
+	// TODO SMOD
+	case evmasm::Instruction::LT:
+		if (m_useInt)
+			return smtutil::Expression::ite(arguments.at(0) < arguments.at(1), constantValue(1), constantValue(0));
```

Comment:
Maybe use helper for all the comparison functions?

Reply:
I didn't know whether to go for virtual helpers for each function/instruction or for the entire `encode` function. I went for the latter, but open for change.